helpers/controllers/authController.test.js

```diff
@@ -81,4 +89,75 @@ describe('auth controller', () => {
       success: true
     })
   })
+
+  test('Signup - should resolve with success false if userinfo is incomplete', async () => {
+    const result = await signup({}, {}, { req: { session: {} } })
+    expect(result).toEqual({
+      success: false,
+      error: 'Invalid registration information'
+    })
+  })
+
+  test('Signup - should resolve with success false if user already exists', async () => {
+    db.User.findOne = jest.fn().mockReturnValue({ username: 'c0d3user' })
+    const result = await signup({}, userArgs, { req: { session: {} } })
+    expect(result).toEqual({
+      success: false,
+      error: 'User already exists'
+    })
+  })
+
+  test('Signup - should resolve with success false if email already exists', async () => {
+    db.User.findOne = jest
+      .fn()
+      .mockReturnValueOnce(null)
+      .mockReturnValue({ username: 'c0d3user' }) // Second call for User.findOne checks for email
+    const result = await signup({}, userArgs, { req: { session: {} } })
+    expect(result).toEqual({
+      success: false,
+      error: 'Email already exists'
+    })
+  })
+
+  test('Signup - should not create user if chat signup response with 401 or 403', async () => {
+    db.User.findOne = jest.fn().mockReturnValue(null)
+    db.User.create = jest.fn()
+
+    chatSignUp.mockResolvedValueOnce({
+      success: false,
+      error: 'Mattermost signup error'
+    })
+
+    chatSignUp.mockRejectedValueOnce({
```

Comment:
Why do we have to mock both the return and rejected value?

Reply:
Shouldn't it either resolve or reject, not both?

Reply:
reject I think may be caused by error (like request network failed), and success false means there may be an validation error that mattermost sent back.